uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h

```diff
@@ -0,0 +1,795 @@
+  // returns a sequence, it'll be lifted into an `nsTArray<T>`, not a
+  // `dom::Sequence<T>`. See the note about sequences above.
+  static bool ReadFrom(Reader& aReader, dom::Sequence<T>& aValue) = delete;
+
+  static void WriteInto(Writer& aWriter, const dom::Sequence<T>& aValue) {
+    SequenceTraits<dom::Sequence<T>>::WriteInto(aWriter, aValue);
+  }
+};
+
+template <typename T>
+struct Serializable<nsTArray<T>> {
+  static size_t Size(const nsTArray<T>& aValue) {
+    return SequenceTraits<nsTArray<T>>::Size(aValue);
+  }
+
+  static bool ReadFrom(Reader& aReader, nsTArray<T>& aValue) {
+    uint32_t length = aReader.ReadUInt32();
+    aValue.SetCapacity(length);
+    aValue.TruncateLength(0);
+    for (uint32_t i = 0; i < length; ++i) {
+      if (!Serializable<T>::ReadFrom(aReader, *aValue.AppendElement())) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  static void WriteInto(Writer& aWriter, const nsTArray<T>& aValue) {
+    SequenceTraits<nsTArray<T>>::WriteInto(aWriter, aValue);
+  }
+};
+
+template <typename K, typename V>
+struct Serializable<Record<K, V>> {
+  static size_t Size(const Record<K, V>& aValue) {
+    CheckedInt<size_t> size;
+    size += sizeof(uint32_t);  // For the length prefix.
+    for (const typename Record<K, V>::EntryType& entry : aValue.Entries()) {
+      size += Serializable<K>::Size(entry.mKey);
+      size += Serializable<V>::Size(entry.mValue);
+    }
+    MOZ_RELEASE_ASSERT(size.isValid());
+    return size.value();
+  }
+
+  static bool ReadFrom(Reader& aReader, Record<K, V>& aValue) {
+    uint32_t length = aReader.ReadUInt32();
+    aValue.Entries().SetCapacity(length);
+    aValue.Entries().TruncateLength(0);
+    for (uint32_t i = 0; i < length; ++i) {
+      typename Record<K, V>::EntryType* entry =
+          aValue.Entries().AppendElement();
+      if (!Serializable<K>::ReadFrom(aReader, entry->mKey)) {
+        return false;
+      }
+      if (!Serializable<V>::ReadFrom(aReader, entry->mValue)) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  static void WriteInto(Writer& aWriter, const Record<K, V>& aValue) {
+    aWriter.WriteUInt32(aValue.Entries().Length());
+    for (const typename Record<K, V>::EntryType& entry : aValue.Entries()) {
+      Serializable<K>::WriteInto(aWriter, entry.mKey);
+      Serializable<V>::WriteInto(aWriter, entry.mValue);
+    }
+  }
+};
+
+/// Partial specialization for all types that can be serialized into a byte
+/// buffer. This is analogous to the `ViaFfiUsingByteBuffer` trait in Rust.
+
+template <typename T>
+struct ViaFfi<T, RustBuffer> {
+  static bool Lift(const RustBuffer& aLowered, T& aLifted) {
+    auto reader = Reader(aLowered);
+    if (!Serializable<T>::ReadFrom(reader, aLifted)) {
+      return false;
+    }
+    if (reader.HasRemaining()) {
+      MOZ_ASSERT(false);
+      return false;
+    }
+    {{ ci.ffi_bytebuffer_free().name() }}(aLowered);
```

Comment:
Freeing a `const RustBuffer&` seems sketchy, even though it’s not actually `const`. Since we’re not inheriting from `Serializable`, just specializing, we can probably change this signature to `Lift(RustBuffer&)`, and the compiler will pick this overload instead of the deleted function in the base template.